### PR TITLE
Add kubebuilder annotations and fix model descriptions

### PR DESCRIPTION
**Pull request by spolti**
Fixes https://github.com/serverlessworkflow/sdk-go/issues/158

**Many thanks for submitting your Pull Request :heart:!**

**What this PR does / why we need it**:

**Special notes for reviewers**:

**Additional information (if needed):**

### CODE REVIEW

**Comment by spolti** [comment]
@ricardozanini it is ready for review.
Added somo TODOs where it needs to have some exclusive validation according to the spec.

File: model/workflow.go [hunk 1]
```diff
@@ -426,9 +480,11 @@ func (s *Schedule) UnmarshalJSON(data []byte) error {
...
 // Cron ...
 type Cron struct {
-	// Repeating interval (cron expression) describing when the workflow instance should be created
+	// Cron expression describing when the workflow instance should be created (automatically).
+	// +kubebuilder:validation:Required
 	Expression string `json:"expression" validate:"required"`
-	// Specific date and time (ISO 8601 format) when the cron expression invocation is no longer valid
+	// Specific date and time (ISO 8601 format) when the cron expression is no longer valid.
+	// +optional
 	ValidUntil string `json:"validUntil,omitempty"`
```
Comment (ricardozanini) [inline model/workflow.go:488]: Can we add the iso8601 validation here too?

File: model/workflow_ref.go [hunk 2]
```diff
@@ -23,16 +23,21 @@ import (
...
+	// +optional
 	Invoke InvokeKind `json:"invoke,omitempty" validate:"required,oneof=async sync"`
-
-	// OnParantComplete specifies how subflow execution should behave when parent workflow completes if invoke is 'async'。
-	// Defaults to terminate.
+	// onParentComplete specifies how subflow execution should behave when parent workflow completes if invoke
+	// is 'async'. Defaults to terminate.
+	// +kubebuilder:validation:Enum=terminate;continue
+	// +kubebuilder:default=terminate
 	OnParentComplete string `json:"onParentComplete,omitempty" validate:"required,oneof=terminate continue"`
```
Comment (ricardozanini) [inline model/workflow_ref.go:41]: Can we have an enum?

File: model/workflow.go [hunk 3]
```diff
@@ -73,42 +73,66 @@ func continueAsStructLevelValidation(structLevel validator.StructLevel) {
...
 	Secrets Secrets `json:"secrets,omitempty"`
-	// Constants Workflow constants are used to define static, and immutable, data which is available to Workflow Expressions.
+	// Constants Workflow constants are used to define static, and immutable, data which is available to
+	// Workflow Expressions.
+	// +optional
 	Constants *Constants `json:"constants,omitempty"`
-	// Identifies the expression language used for workflow expressions. Default is 'jq'
+	// Identifies the expression language used for workflow expressions. Default is 'jq'.
+	// +optional
 	ExpressionLang string `json:"expressionLang,omitempty" validate:"omitempty,min=1"`
```
Comment (ricardozanini) [inline model/workflow.go:115]: I think the spec supports only jq and jsonpath. This can also be an enum.

File: model/foreach_state.go [hunk 4]
```diff
@@ -43,19 +43,26 @@ const (
...
-	// Actions to be executed for each of the elements of inputCollection
+	// actions to be executed for each of the elements of inputCollection
+	// +kubebuilder:validation:MinItems=1
 	Actions []Action `json:"actions,omitempty" validate:"required,min=1,dive"`
-	// State specific timeout
+	// state specific timeout.
 	Timeouts *ForEachStateTimeout `json:"timeouts,omitempty"`
-	// Mode Specifies how iterations are to be performed (sequential or in parallel), defaults to parallel
+	// Specifies how iterations are to be performed (sequential or in parallel), defaults to parallel
 	Mode ForEachModeType `json:"mode,omitempty"`
```
Comment (ricardozanini) [inline model/foreach_state.go:66]: Can we have an enum here? I think we are missing the annotations in the Timeouts too.

File: model/event_state.go [hunk 5]
```diff
@@ -19,17 +19,19 @@ import (
...
+// workflow starting state, the event state definition controls when the workflow instances should be created.
 type EventState struct {
 	// TODO: EventState doesn't have usedForCompensation field.
 
-	// If true consuming one of the defined events causes its associated actions to be performed.
-	// If false all the defined events must be consumed in order for actions to be performed
-	// Defaults to true
+	// If true consuming one of the defined events causes its associated actions to be performed. If false all
+	// the defined events must be consumed in order for actions to be performed. Defaults to true.
 	Exclusive bool `json:"exclusive,omitempty"`
```
Comment (ricardozanini) [inline model/event_state.go:29]: Are we missing the annotations here?

File: model/callback_state.go [hunk 6]
```diff
@@ -18,14 +18,16 @@ import (
...
 	Action Action `json:"action" validate:"required"`
-	// References a unique callback event name in the defined workflow events
+	// References a unique callback event name in the defined workflow events.
+	// +kubebuilder:validation:Required
 	EventRef string `json:"eventRef" validate:"required"`
 	// Time period to wait for incoming events (ISO 8601 format)
+	// +optional
 	Timeouts *CallbackStateTimeout `json:"timeouts,omitempty"`
 	// Event data filter
 	EventDataFilter *EventDataFilter `json:"eventDataFilter,omitempty"`
```
Comment (ricardozanini) [inline model/callback_state.go:33]: Missing kubebuilder

**Review by ricardozanini — COMMENTED** [review]
A few comments only! I think we need to have a way to verify if these annotations are generating the right OpenAPI and validate it in the Kubernetes API to test if it's accepted.

**Comment by spolti** [comment]
> A few comments only! I think we need to have a way to verify if these annotations are generating the right OpenAPI and validate it in the Kubernetes API to test if it's accepted.

indeed, we can try to use kubectl or something else.

Issue created: https://github.com/serverlessworkflow/sdk-go/issues/161

File: model/workflow.go [hunk 7]
```diff
@@ -73,42 +73,66 @@ func continueAsStructLevelValidation(structLevel validator.StructLevel) {
...
+	// +optional
 	AutoRetries bool `json:"autoRetries,omitempty"`
-	// Auth definitions can be used to define authentication information that should be applied to resources defined in the operation
-	// property of function definitions. It is not used as authentication information for the function invocation,
-	// but just to access the resource containing the function invocation information.
+	// Auth definitions can be used to define authentication information that should be applied to resources defined
+	// in the operation property of function definitions. It is not used as authentication information for the
+	// function invocation, but just to access the resource containing the function invocation information.
+	// +optional
 	Auth AuthArray `json:"auth,omitempty" validate:"omitempty"`
```
[the remark this answers is not in the document]
Comment (spolti) [reply model/workflow.go:136]: I don't remember why exactly we did it this way, but we can revisit it later.
Created https://github.com/serverlessworkflow/sdk-go/issues/162

File: model/function.go [hunk 8]
```diff
@@ -44,9 +44,13 @@ type FunctionType string
...
+	// +kubebuilder:validation:Required
 	Name string `json:"name" validate:"required"`
-	// If type is `rest`, <path_to_openapi_definition>#<operation_id>. If type is `rpc`, <path_to_grpc_proto_file>#<service_name>#<service_method>.
-	// If type is `expression`, defines the workflow expression. If the type is `custom`, <path_to_custom_script>#<custom_service_method>.
+	// If type is `rest`, <path_to_openapi_definition>#<operation_id>. If type is `rpc`,
+	// <path_to_grpc_proto_file>#<service_name>#<service_method>.
+	// If type is `expression`, defines the workflow expression. If the type is `custom`,
+	// <path_to_custom_script>#<custom_service_method>.
+	// +kubebuilder:validation:Required
 	Operation string `json:"operation" validate:"required"`
```
[the remark this answers is not in the document]
Comment (ricardozanini) [reply model/function.go:55]: I meant in the `type`. `Operation` is a free string, but required.

File: model/workflow.go [hunk 9]
```diff
@@ -73,42 +73,68 @@ func continueAsStructLevelValidation(structLevel validator.StructLevel) {
...
+	// +optional
 	Constants *Constants `json:"constants,omitempty"`
-	// Identifies the expression language used for workflow expressions. Default is 'jq'
-	ExpressionLang string `json:"expressionLang,omitempty" validate:"omitempty,min=1"`
-	// Timeouts definition for Workflow, State, Action, Branch, and Event consumption.
+	// Identifies the expression language used for workflow expressions. Default is 'jq'.
+	// +kubebuilder:validation:Enum=jq;jsonpath
+	// +kubebuilder:default=jq
+	// +optional
+	ExpressionLang string `json:"expressionLang,omitempty" validate:"omitempty,min=1,oneof=jq jsonpath"`
```
Comment (ricardozanini) [inline model/workflow.go:117]:
```suggestion
	ExpressionLang ExpressionLangType `json:"expressionLang,omitempty" validate:"omitempty,min=1,oneof=jq jsonpath"`
```

Comment (spolti) [reply model/workflow.go:117]: created ExpressionLangType type

File: model/function.go [hunk 10]
```diff
@@ -44,12 +44,21 @@ type FunctionType string
...
 	// Unique function name
+	// +kubebuilder:validation:Required
 	Name string `json:"name" validate:"required"`
-	// If type is `rest`, <path_to_openapi_definition>#<operation_id>. If type is `rpc`, <path_to_grpc_proto_file>#<service_name>#<service_method>.
-	// If type is `expression`, defines the workflow expression. If the type is `custom`, <path_to_custom_script>#<custom_service_method>.
+	// If type is `rest`, <path_to_openapi_definition>#<operation_id>.
+	// If type is `rpc`, <path_to_grpc_proto_file>#<service_name>#<service_method>.
+	// If type is `expression`, defines the workflow expression. If the type is `custom`,
+	// <path_to_custom_script>#<custom_service_method>.
+	// +kubebuilder:validation:Enum=rest;rpc;expression
```
[the remark this answers is not in the document]
Comment (spolti) [reply model/function.go:53]: You've suggested adding previously, if the allowed types are only these three, maybe it does make sense to keep.

[the remark this answers is not in the document]
Comment (ricardozanini) [reply model/function.go:53]: Yes, it was my bad. It was for the `type`. Operation is a free string attribute.

**Pull request merged**